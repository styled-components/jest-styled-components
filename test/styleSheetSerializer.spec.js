import { render } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import styled, { ThemeContext, ThemeProvider } from 'styled-components';

const toMatchSnapshot = component => {
  expect(renderer.create(component).toJSON()).toMatchSnapshot('react-test-renderer');
  expect(shallow(component)).toMatchSnapshot('shallow');
  expect(mount(component)).toMatchSnapshot('mount');
  expect(render(component).container.firstChild).toMatchSnapshot('react-testing-library');
};

const shallowWithTheme = (tree, theme) => {
  // this is terrible but couldn't figure out how to do it properly within the framework of enzyme
  ThemeContext._currentValue = theme;

  return shallow(tree);
};

it('null', () => {
  expect(null).toMatchSnapshot();
});

it('non-styled', () => {
  toMatchSnapshot(<div />);
});

it('empty style', () => {
  const Component = styled.div``;

  toMatchSnapshot(<Component />);
});

it('duplicated components', () => {
  const A = styled.div`
    color: red;
  `;
  const B = styled.div`
    color: green;
  `;

  toMatchSnapshot(
    <div>
      <A /> <A /> <B />
    </div>
  );
});

it('basic', () => {
  const Wrapper = styled.section`
    padding: 4em;
    background: papayawhip;
  `;

  const Title = styled.h1`
    font-size: 1.5em;
    text-align: center;
    color: palevioletred;
  `;

  toMatchSnapshot(
    <Wrapper>
      <Title>Hello World, this is my first styled component!</Title>
    </Wrapper>
  );
});

it('any component', () => {
  const Link = ({ className, children }) => <a className={className}>{children}</a>;

  const StyledLink = styled(Link)`
    color: palevioletred;
    font-weight: bold;
  `;

  toMatchSnapshot(
    <div>
      <Link>Unstyled, boring Link</Link>
      <br />
      <StyledLink>Styled, exciting Link</StyledLink>
    </div>
  );
});

it('attaching additional props', () => {
  const Div = styled.div.attrs(() => ({
    className: 'div',
  }))`
    color: red;
  `;

  toMatchSnapshot(<Div />);
});

it('leading white spaces', () => {
  const Div = () => <div className="  div" />;

  toMatchSnapshot(<Div />);
});

it('trailing white spaces', () => {
  const Div = styled.div.attrs(() => ({
    className: 'div  ',
  }))`
    color: red;
  `;

  toMatchSnapshot(<Div />);
});

it('included class name', () => {
  const Div = styled.div.attrs(() => ({
    className: 'i',
  }))`
    color: red;
  `;

  toMatchSnapshot(<Div />);
});

it('theming', () => {
  const Button = styled.button`
    font-size: 1em;
    margin: 1em;
    padding: 0.25em 1em;
    border-radius: 3px;

    color: ${props => props.theme.main};
    border: 2px solid ${props => props.theme.main};
  `;

  Button.defaultProps = {
    theme: {
      main: 'palevioletred',
    },
  };

  const theme = {
    main: 'mediumseagreen',
  };

  toMatchSnapshot(
    <div>
      <Button>Normal</Button>
      <ThemeProvider theme={theme}>
        <Button>Themed</Button>
      </ThemeProvider>
    </div>
  );
});

it('shallow with theme', () => {
  const Button = styled.button`
    color: ${props => props.theme.main};
  `;

  const theme = {
    main: 'mediumseagreen',
  };

  const wrapper = shallowWithTheme(<Button>Themed</Button>, theme);

  expect(wrapper).toMatchSnapshot();
});

it('supported css', () => {
  const Example = styled.div`
    padding: 2em 1em;
    background: papayawhip;

    &:hover {
      background: palevioletred;
    }

    @media (max-width: 600px) {
      background: tomato;

      &:hover {
        background: yellow;
      }
    }

    > p {
      text-decoration: underline;
    }

    html.test & {
      display: none;
    }
  `;

  toMatchSnapshot(
    <Example>
      <p>Hello World!</p>
    </Example>
  );
});

it('referring to other components', () => {
  const Link = styled.a`
    display: flex;
    align-items: center;
    padding: 5px 10px;
    background: papayawhip;
    color: palevioletred;
  `;

  const Icon = styled.svg`
    transition: fill 0.25s;
    width: 48px;
    height: 48px;

    ${Link}:hover & {
      fill: rebeccapurple;
    }
  `;

  const Label = styled.span`
    display: flex;
    align-items: center;
    line-height: 1.2;

    &::before {
      content: 'â—€';
      margin: 0 10px;
    }
  `;

  const Container = styled.div`
    color: black;
  `;
  const TextWithConditionalFormatting = styled.span`
    ${Container} & {
      color: yellow;
      background-color: ${props => (props.error ? 'red' : 'green')};
    }
  `;

  const component = (
    <Link href="#">
      <Icon />
      <Label>Hovering my parent changes my style!</Label>
      <TextWithConditionalFormatting>I should be green</TextWithConditionalFormatting>
      <TextWithConditionalFormatting error>I should be red</TextWithConditionalFormatting>
    </Link>
  );

  expect(renderer.create(component).toJSON()).toMatchSnapshot('react-test-renderer');
  expect(mount(component)).toMatchSnapshot('mount');
  expect(render(component).container.firstChild).toMatchSnapshot('react-testing-library');
});
