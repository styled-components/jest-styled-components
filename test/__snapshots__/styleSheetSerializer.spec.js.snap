// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`any component: mount 1`] = `
.c0 {
  color: palevioletred;
  font-weight: bold;
}

<div>
  <Link>
    <a>
      Unstyled, boring Link
    </a>
  </Link>
  <br />
  <Styled(Link)>
    <Link
      className="c0"
    >
      <a
        className="c0"
      >
        Styled, exciting Link
      </a>
    </Link>
  </Styled(Link)>
</div>
`;

exports[`any component: react-test-renderer 1`] = `
.c0 {
  color: palevioletred;
  font-weight: bold;
}

<div>
  <a
    className={undefined}
  >
    Unstyled, boring Link
  </a>
  <br />
  <a
    className="c0"
  >
    Styled, exciting Link
  </a>
</div>
`;

exports[`any component: shallow 1`] = `
<div>
  <Link>
    Unstyled, boring Link
  </Link>
  <br />
  <Styled(Link)>
    Styled, exciting Link
  </Styled(Link)>
</div>
`;

exports[`attaching additional props: mount 1`] = `
.c0 {
  color: red;
}

<styled.div>
  <div
    className="div c0"
  />
</styled.div>
`;

exports[`attaching additional props: react-test-renderer 1`] = `
.c0 {
  color: red;
}

<div
  className="div c0"
/>
`;

exports[`attaching additional props: shallow 1`] = `
.c0 {
  color: red;
}

<div
  className="div c0"
/>
`;

exports[`basic: mount 1`] = `
.c0 {
  padding: 4em;
  background: papayawhip;
}

.c1 {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<styled.section>
  <section
    className="c0"
  >
    <styled.h1>
      <h1
        className="c1"
      >
        Hello World, this is my first styled component!
      </h1>
    </styled.h1>
  </section>
</styled.section>
`;

exports[`basic: react-test-renderer 1`] = `
.c0 {
  padding: 4em;
  background: papayawhip;
}

.c1 {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  className="c0"
>
  <h1
    className="c1"
  >
    Hello World, this is my first styled component!
  </h1>
</section>
`;

exports[`basic: shallow 1`] = `
.c0 {
  padding: 4em;
  background: papayawhip;
}

<section
  className="c0"
>
  <styled.h1>
    Hello World, this is my first styled component!
  </styled.h1>
</section>
`;

exports[`duplicated components: mount 1`] = `
.c0 {
  color: red;
}

.c1 {
  color: green;
}

<div>
  <styled.div>
    <div
      className="c0"
    />
  </styled.div>
   
  <styled.div>
    <div
      className="c0"
    />
  </styled.div>
   
  <styled.div>
    <div
      className="c1"
    />
  </styled.div>
</div>
`;

exports[`duplicated components: react-test-renderer 1`] = `
.c0 {
  color: red;
}

.c1 {
  color: green;
}

<div>
  <div
    className="c0"
  />
   
  <div
    className="c0"
  />
   
  <div
    className="c1"
  />
</div>
`;

exports[`duplicated components: shallow 1`] = `
<div>
  <styled.div />
   
  <styled.div />
   
  <styled.div />
</div>
`;

exports[`empty style: mount 1`] = `
<styled.div>
  <div
    className=""
  />
</styled.div>
`;

exports[`empty style: react-test-renderer 1`] = `
<div
  className=""
/>
`;

exports[`empty style: shallow 1`] = `
<div
  className=""
/>
`;

exports[`extending styles: mount 1`] = `
.c0 {
  color: palevioletred;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid palevioletred;
  border-radius: 3px;
}

.c1 {
  color: palevioletred;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid palevioletred;
  border-radius: 3px;
  color: tomato;
  border-color: tomato;
}

<div>
  <styled.button>
    <button
      className="c0"
    >
      Normal Button
    </button>
  </styled.button>
  <styled.button>
    <button
      className="c1"
    >
      Tomato Button
    </button>
  </styled.button>
</div>
`;

exports[`extending styles: react-test-renderer 1`] = `
.c0 {
  color: palevioletred;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid palevioletred;
  border-radius: 3px;
}

.c1 {
  color: palevioletred;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid palevioletred;
  border-radius: 3px;
  color: tomato;
  border-color: tomato;
}

<div>
  <button
    className="c0"
  >
    Normal Button
  </button>
  <button
    className="c1"
  >
    Tomato Button
  </button>
</div>
`;

exports[`extending styles: shallow 1`] = `
<div>
  <styled.button>
    Normal Button
  </styled.button>
  <styled.button>
    Tomato Button
  </styled.button>
</div>
`;

exports[`included class name: mount 1`] = `
.c0 {
  color: red;
}

<styled.div>
  <div
    className="i c0"
  />
</styled.div>
`;

exports[`included class name: react-test-renderer 1`] = `
.c0 {
  color: red;
}

<div
  className="i c0"
/>
`;

exports[`included class name: shallow 1`] = `
.c0 {
  color: red;
}

<div
  className="i c0"
/>
`;

exports[`leading white spaces: mount 1`] = `
<Div>
  <div
    className="  div"
  />
</Div>
`;

exports[`leading white spaces: react-test-renderer 1`] = `
<div
  className="  div"
/>
`;

exports[`leading white spaces: shallow 1`] = `
<div
  className="  div"
/>
`;

exports[`non-styled: mount 1`] = `<div />`;

exports[`non-styled: react-test-renderer 1`] = `<div />`;

exports[`non-styled: shallow 1`] = `<div />`;

exports[`null 1`] = `null`;

exports[`referring to other components: mount 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 10px;
  background: papayawhip;
  color: palevioletred;
}

.c2 {
  -webkit-transition: fill 0.25s;
  transition: fill 0.25s;
  width: 48px;
  height: 48px;
}

.c0:hover .c2 {
  fill: rebeccapurple;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 1.2;
}

.c3::before {
  content: '◀';
  margin: 0 10px;
}

<styled.a
  href="#"
>
  <a
    className="c0 c1"
    href="#"
  >
    <styled.svg>
      <svg
        className="c2"
      />
    </styled.svg>
    <styled.span>
      <span
        className="c3"
      >
        Hovering my parent changes my style!
      </span>
    </styled.span>
  </a>
</styled.a>
`;

exports[`referring to other components: react-test-renderer 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5px 10px;
  background: papayawhip;
  color: palevioletred;
}

.c2 {
  -webkit-transition: fill 0.25s;
  transition: fill 0.25s;
  width: 48px;
  height: 48px;
}

.c0:hover .c2 {
  fill: rebeccapurple;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 1.2;
}

.c3::before {
  content: '◀';
  margin: 0 10px;
}

<a
  className="c0 c1"
  href="#"
>
  <svg
    className="c2"
  />
  <span
    className="c3"
  >
    Hovering my parent changes my style!
  </span>
</a>
`;

exports[`shallow with theme 1`] = `
.c0 {
  color: mediumseagreen;
}

<button
  className="c0"
>
  Themed
</button>
`;

exports[`supported css: mount 1`] = `
.c0 {
  padding: 2em 1em;
  background: papayawhip;
}

.c0:hover {
  background: palevioletred;
}

.c0 > p {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

html.test .c0 {
  display: none;
}

@media (max-width: 600px) {
  .c0 {
    background: tomato;
  }

  .c0:hover {
    background: yellow;
  }
}

<styled.div>
  <div
    className="c0"
  >
    <p>
      Hello World!
    </p>
  </div>
</styled.div>
`;

exports[`supported css: react-test-renderer 1`] = `
.c0 {
  padding: 2em 1em;
  background: papayawhip;
}

.c0:hover {
  background: palevioletred;
}

.c0 > p {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

html.test .c0 {
  display: none;
}

@media (max-width: 600px) {
  .c0 {
    background: tomato;
  }

  .c0:hover {
    background: yellow;
  }
}

<div
  className="c0"
>
  <p>
    Hello World!
  </p>
</div>
`;

exports[`supported css: shallow 1`] = `
.c0 {
  padding: 2em 1em;
  background: papayawhip;
}

.c0:hover {
  background: palevioletred;
}

.c0 > p {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

html.test .c0 {
  display: none;
}

@media (max-width: 600px) {
  .c0 {
    background: tomato;
  }

  .c0:hover {
    background: yellow;
  }
}

<div
  className="c0"
>
  <p>
    Hello World!
  </p>
</div>
`;

exports[`theming: mount 1`] = `
.c0 {
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border-radius: 3px;
  color: palevioletred;
  border: 2px solid palevioletred;
}

.c1 {
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border-radius: 3px;
  color: mediumseagreen;
  border: 2px solid mediumseagreen;
}

<div>
  <styled.button
    theme={
      Object {
        "main": "palevioletred",
      }
    }
  >
    <button
      className="c0"
    >
      Normal
    </button>
  </styled.button>
  <ThemeProvider
    theme={
      Object {
        "main": "mediumseagreen",
      }
    }
  >
    <styled.button
      theme={
        Object {
          "main": "palevioletred",
        }
      }
    >
      <button
        className="c1"
      >
        Themed
      </button>
    </styled.button>
  </ThemeProvider>
</div>
`;

exports[`theming: react-test-renderer 1`] = `
.c0 {
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border-radius: 3px;
  color: palevioletred;
  border: 2px solid palevioletred;
}

.c1 {
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border-radius: 3px;
  color: mediumseagreen;
  border: 2px solid mediumseagreen;
}

<div>
  <button
    className="c0"
  >
    Normal
  </button>
  <button
    className="c1"
  >
    Themed
  </button>
</div>
`;

exports[`theming: shallow 1`] = `
<div>
  <styled.button
    theme={
      Object {
        "main": "palevioletred",
      }
    }
  >
    Normal
  </styled.button>
  <ThemeProvider
    theme={
      Object {
        "main": "mediumseagreen",
      }
    }
  >
    <styled.button
      theme={
        Object {
          "main": "palevioletred",
        }
      }
    >
      Themed
    </styled.button>
  </ThemeProvider>
</div>
`;

exports[`trailing white spaces: mount 1`] = `
.c0 {
  color: red;
}

<styled.div>
  <div
    className="div   c0"
  />
</styled.div>
`;

exports[`trailing white spaces: react-test-renderer 1`] = `
.c0 {
  color: red;
}

<div
  className="div   c0"
/>
`;

exports[`trailing white spaces: shallow 1`] = `
.c0 {
  color: red;
}

<div
  className="div   c0"
/>
`;
